### ISSUE TITLE
yaml pipeline vs. 2

### ISSUE THREAD

**Issue by LamaAni** [woodpecker-ci/woodpecker#771]
### Clear and concise description of the problem

The current YAML configuration dose not match the standards of yaml. E.g. the list values in the pipeline (and thire parsing in config.go) assume the preservation of key order in the pipeline execution.

### Suggested solution

I propose the following:
1. Follow yaml standards in pipeline files.
2. Change the pipeline to be step like.
3. Remove the clone phase and provide the secure key to all stages, allowing the clone to be just another step (if needed).

To support old versions of the yaml file. We should make sure we to add 
api_version: 1/2 on top.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by anbraten** [comment]
@LamaAni I tried to add your suggested points to #567, could you have a look if I copied them correctly?

**Comment by LamaAni** [comment]
@anbraten Why was this closed?

@6543 The pipeline is a dictionary of items and should be a list. In Yaml, dictionary dose not preserve order.

**Comment by 6543** [comment]
yes that's bit ... *interesting* to solfe ... - I'd like to stick with this format for now as it is easy to remember and if we change it it's clearly a breaking change for no benefit beside spec compliant ... and the only project who has to follow whatefer we define are we selfe - so I would call it a nice to have if we do have a v2 pipeline config format at some point ... - but for now we can just let it stay as is

**Comment by LamaAni** [comment]
Hum.. too bad. I think a overhall now that we know what is needed would produce better results, maybe one that follows Github actions notation.

**Comment by 6543** [comment]
well for github actions i have one word: I personal realty dont like them at all ...

**Comment by silverwind** [comment]
Preservation of dict key order depends on the YAML parser in use. Some do, some don't.

It could be made compatible to accept both a array and dict syntax:

```yaml
pipeline:
  step1:
    image: debian
```

```yaml
pipeline:
  - name: step1
    image: debian
```

What worries me more when I read the pipeline docs is that it seems there is no way to run steps in parallel (e.g. there does not seem to be a DAG under the hood).

**Comment by 6543** [comment]
closed by  #1119
